libs/@local/harpc/service/examples/account.rs

```diff
@@ -0,0 +1,195 @@
+#![feature(never_type)]
+#![expect(
+    dead_code,
+    clippy::unwrap_used,
+    clippy::empty_enum,
+    clippy::todo,
+    reason = "non-working example code"
+)]
+
+use error_stack::Report;
+use frunk::HList;
+use futures::{StreamExt, pin_mut, stream};
+use graph_types::account::AccountId;
+use harpc_net::codec::{Decoder, Encoder};
+use harpc_service::{
+    Service,
+    delegate::ServiceDelegate,
+    metadata::Metadata,
+    procedure::{Procedure, ProcedureIdentifier},
+    role::{Client, ClientSession, Role, Server},
+};
+use harpc_tower::{
+    body::{Body, BodyExt},
+    either::Either,
+    request::Request,
+    response::{Parts, Response},
+};
+use harpc_types::{procedure::ProcedureId, service::ServiceId, version::Version};
+use harpc_wire_protocol::{request::procedure::ProcedureDescriptor, response::kind::ResponseKind};
+
+enum AccountProcedureId {
+    CreateAccount,
+}
+
+impl ProcedureIdentifier for AccountProcedureId {
+    fn from_id(id: ProcedureId) -> Option<Self> {
+        match id.value() {
+            0 => Some(Self::CreateAccount),
+            _ => None,
+        }
+    }
+
+    fn into_id(self) -> ProcedureId {
+        match self {
+            Self::CreateAccount => ProcedureId::new(0),
+        }
+    }
+}
+
+struct Account;
+
+impl Service for Account {
+    type ProcedureId = AccountProcedureId;
+    type Procedures = HList![CreateAccount];
```

Comment:
Why we cannot use tuples here?

Reply:
We could, but I found that a `HList` (or similar h-list implementation) is just a lot easier to work with. With tuples you have some unique problems/constraints, like:

* to implement them you need to use a macro, the code is very repetitive
* you have an upper limit (like 16). You can circumvent that limit by using more traits / a proxy trait, but that is just very messy in general and hard to decipher.

That's why I semi recently-ish pivoted to use `HList` from `frunk` for these kind of things, instead of using tuples. It's a lot easier to write and reason about and the `HList!` and `hlist!` macro make it easy to construct these types. (also the crate itself has some cool things like coproduct, but that's beside the point).

I found the code to just be a lot more readable on the implementation side, with a bit of a negative (because now you're compelled to use a type-level macro) but found that the tradeoffs were good enough, if you disagree please let me know and I'll try to use tuples here.

Reply:
We can stick with HList for now so we can make progress here, but generally, I prefer language-constructs (such as tuples). I don't see why construction of `HList` is easier when in a constant scenario such as defining `Procedure`. In addition, the documentation is much easier to read when looking at tuples and if we use a macro to implement a trait on each tuple elements that's fine. Something along these lines:
```rust
// Potential user-defined implementation if required
impl Trait for CreateAccount {...}

impl<T: Trait> Trait for (T,) { ... }
impl<T: Trait, U: Trait> Trait for (T,U) { ... }
```
while the `Trait` bound might not be needed. 

Let's keep it or now, but let's also consider changing it later. Generally, we don't need to create procedures dynamically I guess? If we build it dynamically we could use `frunk` as a helper to construct a tuple.

Reply:
Yes. The reason I like this approach more is that it allows for a clean separation of concerns.

It allows us to have two different traits, one for just the list, another one for the trait itself and it doesn't require a macro to implement everything.

I also found that during dynamically building up a list it's a bit nicer.

I also fully get your point! In a perfect would we would have variable sized tuples we could implement traits over.

Reply:
> I also fully get your point! In a perfect would we would have variable sized tuples we could implement traits over.

Variadic generics or at least variadic tuples would be a game changer for many things 🚀